addons/im_livechat/static/src/embed/common/livechat_service.js

```diff
@@ -117,153 +67,69 @@ export class LivechatService {
-     */
-    _saveLivechatState(saveData) {
-        const { guest_token } = this.store;
-        if (guest_token) {
-            expirableStorage.setItem(GUEST_TOKEN_STORAGE_KEY, guest_token);
-        }
-        const ONE_DAY_TTL = 60 * 60 * 24;
-        if (this.thread.uuid) {
-            cookie.set(LIVECHAT_UUID_COOKIE, this.thread.uuid, ONE_DAY_TTL);
-        }
-        const persisted = this.state === SESSION_STATE.PERSISTED;
-        expirableStorage.setItem(
-            SAVED_STATE_STORAGE_KEY,
-            JSON.stringify({
-                livechatUserId: this.savedState?.livechatUserId ?? user.userId,
-                persisted,
-                store: persisted ? { "discuss.channel": [{ id: this.thread.id }] } : saveData,
-            }),
-            ONE_DAY_TTL
-        );
-        if (this.thread.livechat_operator_id) {
-            expirableStorage.setItem(
-                OPERATOR_STORAGE_KEY,
-                this.thread.livechat_operator_id.id,
-                ONE_DAY_TTL * 7
-            );
-        }
+    async leave(thread) {
+        await rpc("/im_livechat/visitor_leave_session", { channel_id: thread.id });
     }
 
     /**
      * @param {object} param0
      * @param {boolean} [param0.persist=false]
+     * @param {import("models").Thread} [param0.originThread]
      * @returns {Promise<import("models").Thread>}
      */
-    async _createThread({ persist = false }) {
+    async _createThread({ originThread, persist = false }) {
         const data = await rpc(
             "/im_livechat/get_session",
             {
                 channel_id: this.options.channel_id,
                 anonymous_name: this.options.default_username ?? _t("Visitor"),
-                chatbot_script_id: this.savedState
-                    ? this.thread.chatbot?.script.id
-                    : this.rule.chatbotScript?.id,
+                chatbot_script_id:
+                    originThread?.chatbot?.script.id ??
+                    this.store.livechat_rule?.chatbot_script_id?.id,
                 previous_operator_id: expirableStorage.getItem(OPERATOR_STORAGE_KEY),
                 persisted: persist,
             },
             { shadow: true }
         );
         // clean copy of data for saving in storage, because store insert will add cyclic references
-        const saveData = JSON.parse(JSON.stringify(data));
         const { Thread = [] } = this.store.insert(data);
-        this.thread = Thread[0];
-        if (!this.thread?.livechat_operator_id) {
+        if (Thread.length === 0) {
             this.notificationService.add(_t("No available collaborator, please try again later."));
-            this.leave({ notifyServer: false });
             return;
         }
-        this.state = persist ? SESSION_STATE.PERSISTED : SESSION_STATE.CREATED;
-        this._saveLivechatState(saveData);
-        await Promise.all(this._onStateChangeCallbacks[this.state].map((fn) => fn()));
+        return Thread[0];
```

Comment:
Destruct `Thread[0]` is quite weak. We also have it in fetchMessages and this leads to subtle problem. We made a minor tweak in python to order messages, but ideally the RPC result should have `storeData` and a dedicate key containing thread data that id it, if there's one.

Not a huge deal in this PR, this was already the case.

Reply:
Will improve it in a follow up PR. `get_session` should return `storeData` + `channel_id` to be more robust.